File: src/crypto.rs
//! Cryptographic primitives for envelope encryption
//! 
//! This module provides AES-256-GCM encryption/decryption and HKDF key derivation.

use aes_gcm::{
    aead::{Aead, KeyInit, OsRng},
    Aes256Gcm, Nonce,
};
use hkdf::Hkdf;
use rand::RngCore;
use sha2::Sha256;
use zeroize::{Zeroize, ZeroizeOnDrop};

use crate::error::{EnvelopeError, Result};

/// AES-256 key size in bytes (256 bits)
pub const AES_256_KEY_SIZE: usize = 32;

/// AES-GCM nonce size in bytes (96 bits)
pub const NONCE_SIZE: usize = 12;

/// AES-GCM authentication tag size in bytes (128 bits)
pub const TAG_SIZE: usize = 16;

/// A secure key wrapper that zeroizes memory on drop
#[derive(Clone, Zeroize, ZeroizeOnDrop)]
pub struct SecureKey {
    bytes: Vec<u8>,
}

impl SecureKey {
    /// Create a new SecureKey from raw bytes
    pub fn new(bytes: Vec<u8>) -> Self {
        Self { bytes }
    }

    /// Generate a new random AES-256 key
    pub fn generate() -> Self {
        let mut key = vec![0u8; AES_256_KEY_SIZE];
        OsRng.fill_bytes(&mut key);
        Self { bytes: key }
    }

    /// Get the key bytes (use with caution)
    pub fn as_bytes(&self) -> &[u8] {
        &self.bytes
    }

    /// Get the key length
    pub fn len(&self) -> usize {
        self.bytes.len()
    }

    /// Check if the key is empty
    pub fn is_empty(&self) -> bool {
        self.bytes.is_empty()
    }
}

impl std::fmt::Debug for SecureKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "SecureKey([REDACTED])")
    }
}

/// Encrypted data container with nonce and ciphertext
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct EncryptedData {
    /// The nonce used for encryption (12 bytes for AES-GCM)
    pub nonce: Vec<u8>,
    /// The ciphertext including authentication tag
    pub ciphertext: Vec<u8>,
}

impl EncryptedData {
    /// Create a new EncryptedData container
    pub fn new(nonce: Vec<u8>, ciphertext: Vec<u8>) -> Self {
        Self { nonce, ciphertext }
    }

    /// Encode to base64 for storage
    pub fn to_base64(&self) -> String {
        use base64::{engine::general_purpose::STANDARD, Engine};
        let combined = [self.nonce.as_slice(), self.ciphertext.as_slice()].concat();
        STANDARD.encode(combined)
    }

    /// Decode from base64
    pub fn from_base64(encoded: &str) -> Result<Self> {
        use base64::{engine::general_purpose::STANDARD, Engine};
        let decoded = STANDARD
            .decode(encoded)
            .map_err(|e| EnvelopeError::Crypto(format!("Base64 decode error: {}", e)))?;
        
        if decoded.len() < NONCE_SIZE + TAG_SIZE {
            return Err(EnvelopeError::Crypto("Invalid encrypted data length".into()));
        }

        let (nonce, ciphertext) = decoded.split_at(NONCE_SIZE);
        Ok(Self {
            nonce: nonce.to_vec(),
            ciphertext: ciphertext.to_vec(),
        })
    }
}

/// AES-256-GCM cipher wrapper
pub struct AesGcmCipher;

impl AesGcmCipher {
    /// Encrypt plaintext using AES-256-GCM
    /// 
    /// # Arguments
    /// * `key` - The 256-bit encryption key
    /// * `plaintext` - The data to encrypt
    /// * `aad` - Additional authenticated data (optional context)
    /// 
    /// # Returns
    /// Encrypted data containing nonce and ciphertext with auth tag
    pub fn encrypt(key: &SecureKey, plaintext: &[u8], aad: Option<&[u8]>) -> Result<EncryptedData> {
        if key.len() != AES_256_KEY_SIZE {
            return Err(EnvelopeError::Crypto(format!(
                "Invalid key size: expected {}, got {}",
                AES_256_KEY_SIZE,
                key.len()
            )));
        }

        let cipher = Aes256Gcm::new_from_slice(key.as_bytes())
            .map_err(|e| EnvelopeError::Crypto(format!("Cipher init error: {}", e)))?;

        // Generate random nonce
        let mut nonce_bytes = [0u8; NONCE_SIZE];
        OsRng.fill_bytes(&mut nonce_bytes);
        let nonce = Nonce::from_slice(&nonce_bytes);

        // Encrypt with optional AAD
        let ciphertext = match aad {
            Some(aad_data) => {
                use aes_gcm::aead::Payload;
                cipher
                    .encrypt(nonce, Payload { msg: plaintext, aad: aad_data })
                    .map_err(|e| EnvelopeError::Crypto(format!("Encryption error: {}", e)))?
            }
            None => cipher
                .encrypt(nonce, plaintext)
                .map_err(|e| EnvelopeError::Crypto(format!("Encryption error: {}", e)))?,
        };

        Ok(EncryptedData::new(nonce_bytes.to_vec(), ciphertext))
    }

    /// Decrypt ciphertext using AES-256-GCM
    /// 
    /// # Arguments
    /// * `key` - The 256-bit decryption key
    /// * `encrypted` - The encrypted data (nonce + ciphertext)
    /// * `aad` - Additional authenticated data (must match encryption)
    /// 
    /// # Returns
    /// The decrypted plaintext
    pub fn decrypt(key: &SecureKey, encrypted: &EncryptedData, aad: Option<&[u8]>) -> Result<Vec<u8>> {
        if key.len() != AES_256_KEY_SIZE {
            return Err(EnvelopeError::Crypto(format!(
                "Invalid key size: expected {}, got {}",
                AES_256_KEY_SIZE,
                key.len()
            )));
        }

        if encrypted.nonce.len() != NONCE_SIZE {
            return Err(EnvelopeError::Crypto(format!(
                "Invalid nonce size: expected {}, got {}",
                NONCE_SIZE,
                encrypted.nonce.len()
            )));
        }

        let cipher = Aes256Gcm::new_from_slice(key.as_bytes())
            .map_err(|e| EnvelopeError::Crypto(format!("Cipher init error: {}", e)))?;

        let nonce = Nonce::from_slice(&encrypted.nonce);

        // Decrypt with optional AAD
        let plaintext = match aad {
            Some(aad_data) => {
                use aes_gcm::aead::Payload;
                cipher
                    .decrypt(
                        nonce,
                        Payload {
                            msg: &encrypted.ciphertext,
                            aad: aad_data,
                        },
                    )
                    .map_err(|_| EnvelopeError::Crypto("Decryption failed: authentication error".into()))?
            }
            None => cipher
                .decrypt(nonce, encrypted.ciphertext.as_slice())
                .map_err(|_| EnvelopeError::Crypto("Decryption failed: authentication error".into()))?,
        };

        Ok(plaintext)
    }
}

/// HKDF-SHA256 based key derivation
pub struct KeyDerivation;

impl KeyDerivation {
    /// Derive a key using HKDF-SHA256
    /// 
    /// This is an upgrade from simple HMAC-SHA256 derivation,
    /// providing proper extract-and-expand key derivation.
    /// 
    /// # Arguments
    /// * `master_key` - The input keying material (IKM)
    /// * `salt` - Optional salt value (can be derived from context)
    /// * `info` - Context-specific info string (e.g., "KEK" or data_id)
    /// * `output_len` - Desired output key length
    pub fn derive_key(
        master_key: &SecureKey,
        salt: Option<&[u8]>,
        info: &[u8],
        output_len: usize,
    ) -> Result<SecureKey> {
        let hkdf = Hkdf::<Sha256>::new(salt, master_key.as_bytes());
        
        let mut output = vec![0u8; output_len];
        hkdf.expand(info, &mut output)
            .map_err(|e| EnvelopeError::Crypto(format!("Key derivation error: {}", e)))?;

        Ok(SecureKey::new(output))
    }

    /// Derive a key specifically for a data ID (UUID-based context)
    /// 
    /// This provides backward compatibility concept while using HKDF
    /// instead of raw HMAC-SHA256.
    pub fn derive_key_for_data_id(
        master_key: &SecureKey,
        data_id: &uuid::Uuid,
        purpose: &str,
    ) -> Result<SecureKey> {
        let info = format!("envelope-encryption:{}:{}", purpose, data_id);
        Self::derive_key(master_key, None, info.as_bytes(), AES_256_KEY_SIZE)
    }
}

/// Generate cryptographically secure random bytes
pub fn generate_random_bytes(len: usize) -> Vec<u8> {
    let mut bytes = vec![0u8; len];
    OsRng.fill_bytes(&mut bytes);
    bytes
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encrypt_decrypt() {
        let key = SecureKey::generate();
        let plaintext = b"Hello, Envelope Encryption!";
        
        let encrypted = AesGcmCipher::encrypt(&key, plaintext, None).unwrap();
        let decrypted = AesGcmCipher::decrypt(&key, &encrypted, None).unwrap();
        
        assert_eq!(plaintext.to_vec(), decrypted);
    }

    #[test]
    fn test_encrypt_decrypt_with_aad() {
        let key = SecureKey::generate();
        let plaintext = b"Secret data";
        let aad = b"additional context";
        
        let encrypted = AesGcmCipher::encrypt(&key, plaintext, Some(aad)).unwrap();
        let decrypted = AesGcmCipher::decrypt(&key, &encrypted, Some(aad)).unwrap();
        
        assert_eq!(plaintext.to_vec(), decrypted);
        
        // Should fail with wrong AAD
        let result = AesGcmCipher::decrypt(&key, &encrypted, Some(b"wrong aad"));
        assert!(result.is_err());
    }

    #[test]
    fn test_key_derivation() {
        let master = SecureKey::generate();
        let data_id = uuid::Uuid::new_v4();
        
        let derived1 = KeyDerivation::derive_key_for_data_id(&master, &data_id, "DEK").unwrap();
        let derived2 = KeyDerivation::derive_key_for_data_id(&master, &data_id, "DEK").unwrap();
        
        // Same inputs should produce same output
        assert_eq!(derived1.as_bytes(), derived2.as_bytes());
        
        // Different purpose should produce different key
        let derived3 = KeyDerivation::derive_key_for_data_id(&master, &data_id, "KEK").unwrap();
        assert_ne!(derived1.as_bytes(), derived3.as_bytes());
    }

    #[test]
    fn test_encrypted_data_base64() {
        let key = SecureKey::generate();
        let plaintext = b"Test data for base64";
        
        let encrypted = AesGcmCipher::encrypt(&key, plaintext, None).unwrap();
        let encoded = encrypted.to_base64();
        let decoded = EncryptedData::from_base64(&encoded).unwrap();
        
        assert_eq!(encrypted.nonce, decoded.nonce);
        assert_eq!(encrypted.ciphertext, decoded.ciphertext);
    }
}

File: python/envelope_encryption/crypto.py
"""
Cryptographic primitives for AES-256-GCM envelope encryption.

This module provides:
- SecureKey: Secure key wrapper with automatic zeroization
- EncryptedData: Encrypted payload with nonce and ciphertext
- AesGcmCipher: AES-256-GCM encryption/decryption operations
"""

from __future__ import annotations

import base64
import secrets
from dataclasses import dataclass
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .errors import CryptoError

# Cryptographic constants
AES_256_KEY_SIZE: int = 32  # 256 bits
NONCE_SIZE: int = 12  # 96 bits (standard for AES-GCM)
TAG_SIZE: int = 16  # 128 bits (authentication tag)


class SecureKey:
    """
    Secure key wrapper with automatic memory cleanup on deletion.

    Uses bytearray internally for mutable zeroing in __del__.
    Note: Python's garbage collector doesn't guarantee immediate cleanup,
    so this is best-effort zeroization.
    """

    __slots__ = ("_bytes",)

    def __init__(self, key_bytes: bytes | bytearray) -> None:
        """
        Create a SecureKey from raw bytes.

        Args:
            key_bytes: Raw key material (should be 32 bytes for AES-256)
        """
        if not isinstance(key_bytes, (bytes, bytearray)):
            raise CryptoError("Key must be bytes or bytearray")
        self._bytes = bytearray(key_bytes)

    @classmethod
    def generate(cls) -> SecureKey:
        """Generate a cryptographically secure random 32-byte key."""
        return cls(secrets.token_bytes(AES_256_KEY_SIZE))

    def as_bytes(self) -> bytes:
        """Return key as immutable bytes."""
        return bytes(self._bytes)

    def __len__(self) -> int:
        """Return key length in bytes."""
        return len(self._bytes)

    def __repr__(self) -> str:
        """Redacted representation to prevent accidental key disclosure."""
        return "SecureKey([REDACTED])"

    def __del__(self) -> None:
        """Zero memory on deletion (best-effort)."""
        if hasattr(self, "_bytes"):
            for i in range(len(self._bytes)):
                self._bytes[i] = 0


@dataclass
class EncryptedData:
    """
    Encrypted data container with nonce and ciphertext.

    The ciphertext includes the 16-byte authentication tag appended by AESGCM.
    """

    nonce: bytes  # 12 bytes
    ciphertext: bytes  # Ciphertext + 16-byte auth tag

    def to_aead_blob(self) -> bytes:
        """
        Convert to industry-standard AEAD blob format: nonce || ciphertext || tag.

        For AES-256-GCM encrypting a 32-byte key: 12 + 32 + 16 = 60 bytes total.
        """
        return self.nonce + self.ciphertext

    @classmethod
    def from_aead_blob(cls, blob: bytes) -> EncryptedData:
        """
        Parse from AEAD blob format: nonce || ciphertext || tag.

        Args:
            blob: Raw AEAD blob bytes

        Returns:
            EncryptedData instance

        Raises:
            CryptoError: If blob is too small
        """
        min_size = NONCE_SIZE + TAG_SIZE
        if len(blob) < min_size:
            raise CryptoError(
                f"AEAD blob too small: expected at least {min_size} bytes, got {len(blob)}"
            )
        return cls(nonce=blob[:NONCE_SIZE], ciphertext=blob[NONCE_SIZE:])

    def to_base64(self) -> str:
        """Encode as base64 string."""
        return base64.standard_b64encode(self.to_aead_blob()).decode("ascii")

    @classmethod
    def from_base64(cls, encoded: str) -> EncryptedData:
        """
        Decode from base64 string.

        Args:
            encoded: Base64-encoded AEAD blob

        Returns:
            EncryptedData instance

        Raises:
            CryptoError: If decoding fails or data is invalid
        """
        try:
            decoded = base64.standard_b64decode(encoded)
        except Exception as e:
            raise CryptoError(f"Base64 decode error: {e}")

        min_size = NONCE_SIZE + TAG_SIZE
        if len(decoded) < min_size:
            raise CryptoError("Invalid encrypted data length")

        return cls(nonce=decoded[:NONCE_SIZE], ciphertext=decoded[NONCE_SIZE:])


class AesGcmCipher:
    """
    AES-256-GCM authenticated encryption.

    Provides static methods for encryption and decryption with optional
    Additional Authenticated Data (AAD) for binding.
    """

    @staticmethod
    def encrypt(
        key: SecureKey,
        plaintext: bytes,
        aad: Optional[bytes] = None,
    ) -> EncryptedData:
        """
        Encrypt plaintext with AES-256-GCM.

        Args:
            key: 32-byte encryption key
            plaintext: Data to encrypt
            aad: Optional Additional Authenticated Data for binding

        Returns:
            EncryptedData with nonce and ciphertext (includes auth tag)

        Raises:
            CryptoError: If key size is invalid or encryption fails
        """
        if len(key) != AES_256_KEY_SIZE:
            raise CryptoError(
                f"Invalid key size: expected {AES_256_KEY_SIZE}, got {len(key)}"
            )

        nonce = secrets.token_bytes(NONCE_SIZE)
        aesgcm = AESGCM(key.as_bytes())

        try:
            ciphertext = aesgcm.encrypt(nonce, plaintext, aad)
        except Exception as e:
            raise CryptoError(f"Encryption error: {e}")

        return EncryptedData(nonce=nonce, ciphertext=ciphertext)

    @staticmethod
    def decrypt(
        key: SecureKey,
        encrypted: EncryptedData,
        aad: Optional[bytes] = None,
    ) -> bytes:
        """
        Decrypt ciphertext with AES-256-GCM.

        Args:
            key: 32-byte decryption key
            encrypted: EncryptedData with nonce and ciphertext
            aad: Optional Additional Authenticated Data (must match encryption)

        Returns:
            Decrypted plaintext bytes

        Raises:
            CryptoError: If key/nonce size is invalid or decryption fails
        """
        if len(key) != AES_256_KEY_SIZE:
            raise CryptoError(
                f"Invalid key size: expected {AES_256_KEY_SIZE}, got {len(key)}"
            )

        if len(encrypted.nonce) != NONCE_SIZE:
            raise CryptoError(
                f"Invalid nonce size: expected {NONCE_SIZE}, got {len(encrypted.nonce)}"
            )

        aesgcm = AESGCM(key.as_bytes())

        try:
            return aesgcm.decrypt(encrypted.nonce, encrypted.ciphertext, aad)
        except Exception:
            # Generic error to prevent oracle attacks
            raise CryptoError("Decryption failed")


def generate_random_bytes(length: int) -> bytes:
    """
    Generate cryptographically secure random bytes.

    Args:
        length: Number of bytes to generate

    Returns:
        Random bytes of specified length
    """
    return secrets.token_bytes(length)
